contracts/presets/oracles/ConstrainedOracle.sol

```diff
@@ -0,0 +1,146 @@
+pragma solidity >=0.8.0;
+
+import {BaseOracle} from "./BaseOracle.sol";
+import {InitializeOracleParams} from "../../commons/Types.sol";
+import {IKPIToken} from "../../interfaces/kpi-tokens/IKPIToken.sol";
+import {UNIT, INVALID_ANSWER} from "../../commons/Constants.sol";
+
+enum Constraint {
+    Between,
+    NotBetween,
+    // range is very similar to between, except the final completion
+    // percentage is determined based on the final absolute result
+    // checking where it landed in the specified range. As an example,
+    // if we have a range constraint with bounds 0 and 10 and the end
+    //  result is 5, the goal will be determined to have been 50% completed.
+    Range,
+    GreaterThan,
+    LowerThan,
+    Equal,
+    NotEqual
+}
+
+/// SPDX-License-Identifier: GPL-3.0-or-later
+/// @title ConstrainedOracle contract
+/// @dev An oracle preset that can be initialized with a
+/// constraint on the end result and that transforms the end
+/// result itself in a Carrot-compatible target completion
+/// percentage based on it on finalization.
+/// @author Federico Luzzi - <federico.luzzi@protonmail.com>
+abstract contract ConstrainedOracle is BaseOracle {
+    uint256 internal constant MULTIPLIER = 64;
+
+    Constraint internal constraint;
+    uint256 internal value0;
+    uint256 internal value1;
+
+    error InvalidConstraint();
+    error InvalidValues();
+    error InvalidRangeBounds();
+    error InvalidGreaterThanConstraintValue();
+    error InvalidLowerThanConstraintValue();
+
+    /// @dev Initializes the base oracle preset this contract
+    /// extends from and sets up the constraint on the end result
+    /// of the oracle based on the given parameters. This function
+    ///  can only be called while initializing an oracle instance. If you extend from this contract, make sure you call this, otherwise no state will be initialized.
```

Comment:
Additional space here.

Reply:
Addressed in 085321c61ca8d07a7d23bc66030e6a08ff25b02e